### PR TITLE
Refactor update-dependencies to support additional scenarios

### PR DESCRIPTION
**Pull request by MichaelSimons**
* Add support for reading local build info file.
* Add support for 2.0 sdk scenarios where the sdk version does not align with the dockerfile version folder layout.  Runtime version info is used to find Dockerfiles now.
* Add support for reading checksums from `https://dotnetcli.blob.core.windows.net/dotnet/checksums` if not found in `https://dotnetclichecksums.blob.core.windows.net`

@dotnet-bot skip ci please

### CODE REVIEW

File: scripts/update-dependencies/DockerfileShaUpdater.cs [hunk 1]
```diff
@@ -18,64 +18,136 @@ namespace Dotnet.Docker
...
+            $"ENV (?<{EnvNameGroupName}>(DOTNET|ASPNETCORE)_[\\S]*VERSION) (?<{ValueGroupName}>[\\S]*)";
+        private static readonly string s_envShaPattern =
+            $"ENV (?<{EnvNameGroupName}> DOTNET_[\\S]*DOWNLOAD_SHA) (?<{ValueGroupName}>[\\S]*)";
+        private static readonly string s_envDownloadUrlPattern = $"ENV (DOTNET_[\\S]*DOWNLOAD_URL) (?<{ValueGroupName}>[\\S]*)";
+        private static readonly string s_inlineUrlPattern = 
+            $"(?<{ValueGroupName}>https://dotnetcli.blob.core.windows.net/[^;\\s]*)";
+        private static readonly string s_varShaPattern =
+            $"[ \\$](?<{EnvNameGroupName}>(dotnet_|aspnetcore_)sha512)( )*=( )*'(?<{ValueGroupName}>[^'\\s]*)'";
+
+        private static readonly Regex s_downloadUrlRegex = new Regex($"({s_envDownloadUrlPattern})|{s_inlineUrlPattern}");
```
Comment (dagood) [inline scripts/update-dependencies/DockerfileShaUpdater.cs]: Unfortunate that GitHub's code coloring doesn't like that `s_varShaPattern` string. 😕 Not sure what's exactly causing it to break.

File: scripts/update-dependencies/DockerfileShaUpdater.cs [hunk 2]
```diff
@@ -18,64 +18,136 @@ namespace Dotnet.Docker
...
-                    Trace.TraceInformation($"Downloading '{shaUrl}'.");
-                    using (Stream shaStream = new HttpClient().GetStreamAsync(shaUrl).Result)
-                    using (StreamReader reader = new StreamReader(shaStream))
-                    {
-                        sha = reader.ReadToEnd().ToLowerInvariant();
-                    }
+                    downloadUrl = SubstituteEnvRef(downloadUrl, versionInfo.EnvName, versionInfo.Version);
+                    sha = await GetDotNetCliChecksumsShaAsync(downloadUrl, versionInfo.EnvName)
+                        ?? await GetDotNetReleaseChecksumsShaAsync(downloadUrl, versionInfo.EnvName, versionInfo.Version);
+                    sha = sha?.ToLower();
```
Comment (dagood) [inline scripts/update-dependencies/DockerfileShaUpdater.cs]: `ToLowerInvariant`?

File: scripts/update-dependencies/DockerfileShaUpdater.cs [hunk 3]
```diff
@@ -18,64 +18,136 @@ namespace Dotnet.Docker
...
+            {
+                if (response.IsSuccessStatusCode)
+                {
+                    sha = await response.Content.ReadAsStringAsync();
+                }
+                else
+                {
+                    Trace.TraceInformation($"Failed to find `dotnetclichecksums` sha");
+                }
+            };
```
Comment (dagood) [inline scripts/update-dependencies/DockerfileShaUpdater.cs]: Unnecessary semicolon (and more later on other `using`s).

File: scripts/update-dependencies/Options.cs [hunk 4]
```diff
@@ -42,10 +42,11 @@ public void Parse(string[] args)
                     "GitHub user used to make PR (if not specified, a PR will not be created)");
                 GitHubUser = gitHubUser;
 
-                string buildInfoUrl = null;
+                Uri buildInfoUrl = null;
                 syntax.DefineParameter(
                     "build-info",
                     ref buildInfoUrl,
+                    (value) => new Uri(value),
                     "URL of the build info to update the Dockerfiles with");
```
Comment (dagood) [inline scripts/update-dependencies/Options.cs]: Worthwhile to call out here that `file://` works?

File: scripts/update-dependencies/DockerfileShaUpdater.cs (hunk 2 above)
Comment (MichaelSimons) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: Switched

File: scripts/update-dependencies/DockerfileShaUpdater.cs (hunk 3 above)
Comment (MichaelSimons) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: Removed all

File: scripts/update-dependencies/Options.cs (hunk 4 above)
Comment (MichaelSimons) [reply scripts/update-dependencies/Options.cs]: Good suggestion.

File: scripts/update-dependencies/DockerfileShaUpdater.cs (hunk 1 above)
Comment (MichaelSimons) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: I had noticed that :(

Comment (MichaelSimons) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: If I don't put it on a separate line, the syntax coloring works...

Comment (dagood) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: Wow... that's pretty strange.

Comment (MichaelSimons) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: I wonder if this is by chance related to the same underlying cause as https://github.com/atom/language-csharp/issues/113?  Sounds like work is in play to get GH to use a more recent version of the grammar.  Will monitor.

Comment (kasper3) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: FYI, GitHub has just released and that issue along with many others around C# syntax highlighting are fixed 🎉

Comment (MichaelSimons) [reply scripts/update-dependencies/DockerfileShaUpdater.cs]: Thanks for the heads up @kasper3!

**Pull request merged**